torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2987,6 +2987,22 @@ bool any_variable_defined(variable_list& variables) {
   return false;
 }
 
+std::tuple<Tensor, Tensor> polar_backward(
+    const Tensor& grad,
+    const Tensor& abs,
+    const Tensor& angle) {
+  Tensor grad_abs, grad_angle;
+  if (grad.defined()) {
+    auto angle_cos = angle.cos();
+    auto angle_sin = angle.sin();
+    auto grad_real = at::real(grad);
+    auto grad_imag = at::imag(grad);
+    grad_abs = (grad_real * angle_cos) + (grad_imag * angle_sin);
+    grad_angle = (grad_real * abs * -angle_sin) + (grad_imag * abs * angle_cos);
```

Comment:
I think we could improve the performance here by taking result of the forward as input, and then rewriting this as:
```
auto result_mul_1_j = torch.view_as_real(result * 1j);
auto grad_as_real = torch.view_as_real(grad);
grad_angle = torch.dot(result_mul_1_j, grad_as_real).sum(-1);
grad_abs = torch.dot(torch.view_as_real(torch.sgn(result)), grad_as_real).sum(-1);
```
Trigonometric functions are computationally expensive! and since we already computed them during forward it makes sense to use that result. However, you should verify if the code I suggested does indeed give us a mark up over what you have rn.

Reply:
Ah. That makes sense. 
I played around with the mentioned formula below

<details> 

  <summary> Code </summary>

```python
import torch
import numpy as np

# abs = torch.tensor([1, 2], dtype=torch.float64, requires_grad=True)
# angle = torch.tensor([np.pi / 2, 5 * np.pi / 4], dtype=torch.float64, requires_grad=True)

abs = torch.randn(2, 2, dtype=torch.float64, requires_grad=True)
angle = torch.randn(2, 2, dtype=torch.float64, requires_grad=True)

# Inherit from Function
class Polar(torch.autograd.Function):
    @staticmethod
    def forward(ctx, abs, angle):
        
        with torch.no_grad():
            output = torch.polar(abs, angle)
        ctx.save_for_backward(abs, angle, output)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        abs, angle, result = ctx.saved_tensors
        result_mul_1_j = torch.view_as_real(result * complex(0, 1))
        grad_as_real = torch.view_as_real(grad_output)
        grad_angle = (result_mul_1_j * grad_as_real).sum(-1)
        grad_abs = (torch.view_as_real(torch.sgn(result)) * grad_as_real).sum(-1)
        # grad_abs = (torch.view_as_real(result/abs) * grad_as_real).sum(-1) # If abs has negative values

        return grad_abs, grad_angle

polar = Polar.apply

def fn(abs, angle):
    return polar(abs, angle)

torch.autograd.gradcheck(fn, [abs, angle])
torch.autograd.gradgradcheck(fn, [abs, angle])
```

</details>

It has a problem when `abs` could be negative (which it shouldn't be, but there is no check for it)

```python
>> abs = torch.tensor([-1, -2], dtype=torch.float64, requires_grad=True)
>> angle = torch.tensor([np.pi / 2, 5 * np.pi / 4], dtype=torch.float64, requires_grad=True)
>> torch.polar(abs, angle)
tensor([-6.1232e-17-1.0000j,  1.4142e+00+1.4142j], dtype=torch.complex128,
       grad_fn=<PolarBackward>)
```

In that case, we might want to add a relevant check in the kernel or  update the formula to 
`grad_abs = (torch.view_as_real(result/abs) * grad_as_real).sum(-1) # If abs has negative values`.

I think we can add relevant check here,
https://github.com/pytorch/pytorch/blob/bb34fd6191e0f2a31edc38e00d93a24d9da8c031/aten/src/ATen/native/cpu/ComplexKernel.cpp#L20-L22

I am not sure how to assert in CUDA kernel context.

Let me know what you feel about it.

Reply:
Turns out the test itself checks with negative `abs` values

https://github.com/pytorch/pytorch/blob/bb34fd6191e0f2a31edc38e00d93a24d9da8c031/test/test_tensor_creation_ops.py#L496-L507

Reply:
@anjali411 Gentle Ping :)

Reply:
@kshitij12345 afaik `torch.abs` always returns a positive value for all dtypes.

Reply:
Not `torch.abs` but the `abs` argument of `torch.polar`. From the name `abs` one would expect it to be `+ve` but there are no checks in `torch.polar` to verify the same and test mentioned above tests with `-ve` values for abs.

Function Signature: `torch.polar(abs, angle, *, out=None)`
https://pytorch.org/docs/stable/generated/torch.polar.html#torch.polar

Reply:
oh, my bad! good catch. yeah we should add checks to test that input `abs` is non-negative. But, it might be better to fix that in a separate PR!

Reply:
On more thought, I think we can further simplify the formula:
```
grad_real = torch.real(grad);
grad_angle=-1 * torch.imag(result) * grad_real;
grad_abs = torch.cos(angle) * grad_real;
```
this should be faster than the other implementations we discussed above. could you verify?

Reply:
The formula looks incorrect, for `grad_abs` it should be  `torch.cos(angle) * grad_real + torch.sin(angle) * grad_imag` (gradcheck passes) and `grad_angle` is also incorrect as gradcheck is failing.

Reply:
ahh yeah I messed up. this should be correct I think:
```
grad_angle = -r * (g1 * sin θ + g2 * cos θ)
grad_abs = (g1*cos θ + g2 * sin θ)
```
where `grad=g1+g2j`

Reply:
This is similar to the trignometric version with factoring out the abs in terms of operations.
```
grad_abs = (grad_real * angle_cos) + (grad_imag * angle_sin);
grad_angle = abs * (grad_real * -angle_sin) + (grad_imag * angle_cos);
```

Will try and update.

Reply:
@kshitij12345 just to be clear, the g1 g2 formula I wrote above is just to check the correctness on pen and paper. I am suggesting to try the code I added at the end in that comment.

Reply:
I benchmarked it on CPU.
```
[ polar.backward() input dtype:torch.float64 device:cpu ]
                             |  Using third impl
1 threads: -----------------------------------------
      (128,)                           |             88.9
      (256, 128)                   |           1647.4
      (2, 512, 256)               |          18289.6
      (2, 64, 256, 128)        |         417904.6
      (4, 2, 512, 256, 128)  |       13032172.2

Times are in microseconds (us).
```

Reply:
@anjali411 Thanks for looking into it. Updating the benchmark comment for the same. Looks to be faster in general.